include/dlaf/communication/communicator_grid.h

```diff
@@ -0,0 +1,80 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+#include <array>
+
+#include "dlaf/common/index2d.h"
+#include "dlaf/communication/communicator.h"
+
+namespace dlaf {
+namespace comm {
+
+/// @brief Create a communicator with a 2D Grid structure
+///
+/// Given a communicator, it creates communicators for rows and columns, completely independent from the
+/// original one. These new communicators lifetimes management is up to the CommunicatorGrid.
+///
+/// If the grid size does not cover the entire set of ranks available in the original Communicator,
+/// there will be ranks that will be not part of the row and column communicators. On the opposite,
+/// if a grid size bigger that overfit the available number of ranks is specified, it will raise an
+/// exception.
+///
+/// CommunicatorGrid must be destroyed before calling MPI_Finalize, to allow it releasing resources.
+class CommunicatorGrid {
+  struct TAG_MPI;
+  using IndexType = int;
+
+public:
+  using Index2D = common::Index2D<IndexType, TAG_MPI>;
+  using Size2D = common::Size2D<IndexType, TAG_MPI>;
```

Comment:
I would move the index and the size out of the class.
(Index that are "shared" between different classes should be defined outside the classes, therefore, to be consistent, we define all indices outside classes)